airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/abstract_stream.py

```diff
@@ -0,0 +1,152 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+import logging
+from abc import ABC, abstractmethod
+from functools import lru_cache
+from typing import Any, Iterable, List, Mapping, Optional, Tuple, Union
+
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.core import StreamData
+from airbyte_cdk.sources.utils import casing
+from airbyte_cdk.sources.utils.slice_logger import SliceLogger
+from deprecated.classic import deprecated
+
+
+@deprecated("This class is experimental. Use at your own risk.")
+class AbstractStream(ABC):
+    """
+    AbstractStream is an experimental interface for streams developed as part of the Concurrent CDK.
+    This interface is not yet stable and may change in the future. Use at your own risk.
+
+    Why create a new interface instead of adding concurrency capabilities the existing Stream?
+    We learnt a lot since the initial design of the Stream interface, and we wanted to take the opportunity to improve.
+
+    High level, the changes we are targeting are:
+    - Removing superfluous or leaky parameters from the methods' interfaces
+    - Using composition instead of inheritance to add new capabilities
+
+    To allow us to iterate fast while ensuring backwards compatibility, we are creating a new interface with a facade object that will bridge the old and the new interfaces.
+    Source connectors that which to leverage concurrency need to implement this new interface. An example will be available shortly
```

Comment:
while it's true that we want to migrate connectors to the new interface, we should be able to punt on this by creating an [adapter method to create a concurrent stream from a stream](https://github.com/airbytehq/airbyte/pull/30596/files#diff-e3651091c66723be70e9ab4b702a078c07f0e18fdbf895bba1b287a161576f3fR37-R56)

```
    def create_from_legacy_stream(cls, stream: Stream, source: AbstractSource, max_workers: int) -> Stream:
        """
        Create a ConcurrentStream from a legacy Stream.
        :param source:
        :param stream:
        :param max_workers:
        :return:
        """
        return StreamFacade(
            ConcurrentStream(
                partition_generator=LegacyPartitionGenerator(stream),
                max_workers=max_workers,
                name=stream.name,
                json_schema=stream.get_json_schema(),
                availability_strategy=LegacyAvailabilityStrategy(stream, source),
                primary_key=stream.primary_key,
                cursor_field=stream.cursor_field,
            )
        )
```
The benefit is it would allow us to start using concurrency on stripe without the need to re-write the streams, which might be involved due to the dependency between `AvailabilityStrategy` and `Stream`